File: src/routes/auth.routes.ts
import { Hono } from "hono";
import { AuthService } from "@/services/auth.service";
import { successResponse } from "@/utils/response";
import { z } from "zod";
import { phoneNumberSchema } from "@/types/phone_number";
import type { Env } from "@/types/hono";
import { authMiddleware } from "@/middleware/auth";

export const authRoutes = new Hono<Env>();
const authService = new AuthService();

const registerSchema = z.object({
  token: z.string().min(1, "Token zorunludur"),
  name: z.string().min(2, "İsim en az 2 karakter olmalıdır"),
  surname: z.string().min(2, "Soyisim en az 2 karakter olmalıdır"),
});

authRoutes.post("/register", async (c) => {
  const body = await c.req.json();
  const { token, name, surname } = registerSchema.parse(body);

  const user = await authService.registerUser(token, { name, surname });

  return c.json(successResponse(user), 201);
});

const isRegisteredSchema = z.object({
  phone: phoneNumberSchema,
});

authRoutes.post("/is-registered", async (c) => {
  const body = await c.req.json();
  const { phone } = isRegisteredSchema.parse(body);

  const result = await authService.isNumberRegistered(phone);

  return c.json(successResponse(result));
});

authRoutes.get("/users/:id", async (c) => {
  const userId = c.req.param("id");

  const user = await authService.getUserById(userId);

  return c.json(successResponse(user));
});

const updateProfileSchema = z.object({
  name: z.string().min(2, "İsim en az 2 karakter olmalıdır"),
  surname: z.string().min(2, "Soyisim en az 2 karakter olmalıdır"),
});

authRoutes.put("/me", authMiddleware, async (c) => {
  const user = c.get("user");
  const body = await c.req.json();
  const { name, surname } = updateProfileSchema.parse(body);

  const updated = await authService.updateProfile(user._id!, { name, surname });

  return c.json(successResponse(updated));
});

// DELETE /api/auth/me - Delete account
authRoutes.delete("/me", authMiddleware, async (c) => {
  const user = c.get("user");
  await authService.deleteAccount(user._id!);
  return c.json(successResponse({ success: true, message: "Account deleted successfully" }));
});

File: fisoft/src/main/kotlin/routes/AuthRoutes.kt
package com.ilyasdemirkiran.routes

import com.ilyasdemirkiran.middleware.authenticateUser
import com.ilyasdemirkiran.middleware.currentUser
import com.ilyasdemirkiran.services.AuthService
import com.ilyasdemirkiran.utils.successResponse
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import kotlinx.serialization.Serializable

@Serializable
data class RegisterRequest(
    val token: String,
    val name: String,
    val surname: String
)

@Serializable
data class IsRegisteredRequest(
    val phone: String
)

@Serializable
data class UpdateProfileRequest(
    val name: String,
    val surname: String
)

fun Route.authRoutes() {
    val authService = AuthService()

    route("/api/auth") {
        post("/register") {
            val req = call.receive<RegisterRequest>()
            val user = authService.registerUser(req.token, req.name, req.surname)
            call.respond(HttpStatusCode.Created, successResponse(user))
        }

        post("/is-registered") {
            val req = call.receive<IsRegisteredRequest>()
            val result = authService.isNumberRegistered(req.phone)
            call.respond(HttpStatusCode.OK, successResponse(result))
        }

        get("/users/{id}") {
            val id = call.parameters["id"] ?: ""
            val user = authService.getUserById(id)
            call.respond(HttpStatusCode.OK, successResponse(user))
        }

        put("/me") {
            call.authenticateUser()
            val user = call.currentUser
            val req = call.receive<UpdateProfileRequest>()
            val updated = authService.updateProfile(user._id, req.name, req.surname)
            call.respond(HttpStatusCode.OK, successResponse(updated))
        }

        delete("/me") {
            call.authenticateUser()
            val user = call.currentUser
            authService.deleteAccount(user._id)
            call.respond(HttpStatusCode.OK, successResponse(mapOf("message" to "Account deleted successfully")))
        }
    }
}
